payjoin/src/send/mod.rs

```diff
@@ -239,56 +271,81 @@ impl Configuration {
+    /// contribution is provided then instead of returning error the fee contribution will
+    /// be just lowered in the request to match the change amount.
+    pub fn build_with_additional_fee(
+        mut self,
         max_fee_contribution: bitcoin::Amount,
         change_index: Option<usize>,
-    ) -> Self {
-        Configuration {
-            disable_output_substitution: false,
-            fee_contribution: Some((max_fee_contribution, change_index)),
-            clamp_fee_contribution: false,
-            min_fee_rate: FeeRate::ZERO,
-        }
+        min_fee_rate: FeeRate,
+        clamp_fee_contribution: bool,
+    ) -> Result<(Request, Context), CreateRequestError> {
+        self.fee_contribution = Some((max_fee_contribution, change_index));
+        self.clamp_fee_contribution = clamp_fee_contribution;
+        self.min_fee_rate = min_fee_rate;
+        self.build()
     }
 
     /// Perform Payjoin without incentivizing the payee to cooperate.
     ///
     /// While it's generally better to offer some contribution some users may wish not to.
     /// This function disables contribution.
-    pub fn non_incentivizing() -> Self {
-        Configuration {
-            disable_output_substitution: false,
-            fee_contribution: None,
-            clamp_fee_contribution: false,
-            min_fee_rate: FeeRate::ZERO,
-        }
-    }
-
-    /// Disable output substitution even if the receiver didn't.
-    ///
-    /// This forbids receiver switching output or decreasing amount.
-    /// It is generally **not** recommended to set this as it may prevent the receiver from
-    /// doing advanced operations such as opening LN channels and it also guarantees the
-    /// receiver will **not** reward the sender with a discount.
-    pub fn always_disable_output_substitution(mut self, disable: bool) -> Self {
-        self.disable_output_substitution = disable;
-        self
-    }
-
-    /// Decrease fee contribution instead of erroring.
-    ///
-    /// If this option is set and a transaction with change amount lower than fee
-    /// contribution is provided then instead of returning error the fee contribution will
-    /// be just lowered to match the change amount.
-    pub fn clamp_fee_contribution(mut self, clamp: bool) -> Self {
-        self.clamp_fee_contribution = clamp;
-        self
+    pub fn build_non_incentivizing(mut self) -> Result<(Request, Context), CreateRequestError> {
+        // since this is a builder, these should already be cleared
+        // but we'll reset them to be sure
+        self.fee_contribution = None;
+        self.clamp_fee_contribution = false;
+        self.min_fee_rate = FeeRate::ZERO;
+        self.build()
     }
 
-    /// Sets minimum fee rate required by the sender.
-    pub fn min_fee_rate(mut self, fee_rate: FeeRate) -> Self {
-        self.min_fee_rate = fee_rate;
-        self
+    fn build(self) -> Result<(Request, Context), CreateRequestError> {
```

Comment:
my biggest concern with this is whether or not to mutate self in place or just return what we've got. I think this is OK as is